From 91296fd4fd0f91c74ff90d16f686c55115312c7c Mon Sep 17 00:00:00 2001
From: Binh Nguyen <binh.nguyen.uw@renesas.com>
Date: Thu, 16 Aug 2018 11:43:55 +0700
Subject: [PATCH 624/641] ARM: shmobile: Add resource reset for booting
 secondary CPU

Enable the reset requests by CA7/CA15 debug resource
control register before secondary CPU boot.

Refer to commit 7910fe7dc22e ("ARM: shmobile: r8a7790:....") from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-backport.git

Signed-off-by: Kazumi Harada <kazumi.harada.rh@renesas.com>
Signed-off-by: hienhuynh <hien.huynh.px@rvc.renenas.com>
Signed-off-by: Binh Nguyen <binh.nguyen.uw@renesas.com>
---
 arch/arm/mach-shmobile/smp-r8a7743.c  | 12 ++++++++++++
 arch/arm/mach-shmobile/smp-r8a7744.c  | 12 ++++++++++++
 arch/arm/mach-shmobile/smp-r8a77470.c | 12 ++++++++++++
 3 files changed, 36 insertions(+)

diff --git a/arch/arm/mach-shmobile/smp-r8a7743.c b/arch/arm/mach-shmobile/smp-r8a7743.c
index 3cf4452..d2a4f6e 100644
--- a/arch/arm/mach-shmobile/smp-r8a7743.c
+++ b/arch/arm/mach-shmobile/smp-r8a7743.c
@@ -25,6 +25,9 @@
 #include "r8a7743.h"
 #include "rcar-gen2.h"
 
+#define APMU		0xe6151000
+#define CA15DBGRCR	0x1180
+
 static struct rcar_apmu_config r8a7743_apmu_config[] = {
 	{
 		.iomem = DEFINE_RES_MEM(0xe6152000, 0x188),
@@ -39,6 +42,15 @@ static void __init r8a7743_smp_prepare_cpus(unsigned int max_cpus)
 				       r8a7743_apmu_config,
 				       ARRAY_SIZE(r8a7743_apmu_config));
 
+	/* setup for debug mode */
+	{
+		void __iomem *p = ioremap_nocache(APMU, 0x2000);
+		u32 val = readl_relaxed(p + CA15DBGRCR);
+
+		writel_relaxed((val | 0x01f80000), p + CA15DBGRCR);
+		iounmap(p);
+	}
+
 	rcar_gen2_pm_init();
 }
 
diff --git a/arch/arm/mach-shmobile/smp-r8a7744.c b/arch/arm/mach-shmobile/smp-r8a7744.c
index 483f2cb..a2dca2b 100644
--- a/arch/arm/mach-shmobile/smp-r8a7744.c
+++ b/arch/arm/mach-shmobile/smp-r8a7744.c
@@ -25,6 +25,9 @@
 #include "r8a7744.h"
 #include "rcar-gen2.h"
 
+#define APMU		0xe6151000
+#define CA15DBGRCR	0x1180
+
 static struct rcar_apmu_config r8a7744_apmu_config[] = {
 	{
 		.iomem = DEFINE_RES_MEM(0xe6152000, 0x188),
@@ -39,6 +42,15 @@ static void __init r8a7744_smp_prepare_cpus(unsigned int max_cpus)
 				       r8a7744_apmu_config,
 				       ARRAY_SIZE(r8a7744_apmu_config));
 
+	/* setup for debug mode */
+	{
+		void __iomem *p = ioremap_nocache(APMU, 0x2000);
+		u32 val = readl_relaxed(p + CA15DBGRCR);
+
+		writel_relaxed((val | 0x01f80000), p + CA15DBGRCR);
+		iounmap(p);
+	}
+
 	rcar_gen2_pm_init();
 }
 
diff --git a/arch/arm/mach-shmobile/smp-r8a77470.c b/arch/arm/mach-shmobile/smp-r8a77470.c
index b3ad7c42..061410f 100644
--- a/arch/arm/mach-shmobile/smp-r8a77470.c
+++ b/arch/arm/mach-shmobile/smp-r8a77470.c
@@ -26,6 +26,9 @@
 #include "r8a77470.h"
 #include "rcar-gen2.h"
 
+#define APMU		0xe6151000
+#define CA7DBGRCR	0x0180
+
 static struct rcar_sysc_ch r8a77470_ca7_scu = {
 	.chan_offs = 0x100, /* PWRSR3 .. PWRER3 */
 	.isr_bit = 21, /* CA7-SCU */
@@ -45,6 +48,15 @@ static void __init r8a77470_smp_prepare_cpus(unsigned int max_cpus)
 				       r8a77470_apmu_config,
 				       ARRAY_SIZE(r8a77470_apmu_config));
 
+	/* setup for debug mode */
+	{
+		void __iomem *p = ioremap_nocache(APMU, 0x2000);
+		u32 val = readl_relaxed(p + CA7DBGRCR);
+
+		writel_relaxed((val | 0x01f83330), p + CA7DBGRCR);
+		iounmap(p);
+	}
+
 	rcar_gen2_pm_init();
 	rcar_sysc_power_up(&r8a77470_ca7_scu);
 }
-- 
2.7.4

